File: autocontext/src/autocontext/session/background_session_outcomes.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any, Literal, TypeAlias
from urllib.parse import quote

from autocontext.session.background_session_events import NormalizedSessionEvent, build_artifact_created_session_event
from autocontext.session.background_session_read_model import BackgroundSessionArtifact

SessionOutcomeKind: TypeAlias = Literal[
    "branch",
    "commit",
    "pull_request",
    "screenshot",
    "report",
    "trace",
    "dataset",
    "verification_result",
]
SessionOutcomeStatus: TypeAlias = Literal["available", "pending", "unavailable"]
SessionOutcomeMetadataValue: TypeAlias = str | int | float | bool
SessionOutcome: TypeAlias = dict[str, Any]

_OUTCOME_KINDS: set[str] = {
    "branch",
    "commit",
    "pull_request",
    "screenshot",
    "report",
    "trace",
    "dataset",
    "verification_result",
}
_OUTCOME_STATUSES: set[str] = {"available", "pending", "unavailable"}


def build_session_outcome(
    *,
    session_id: str,
    kind: SessionOutcomeKind,
    created_at: str,
    outcome_id: str | None = None,
    status: SessionOutcomeStatus = "available",
    title: str = "",
    url: str = "",
    path: str = "",
    ref: str = "",
    sha: str = "",
    summary: str = "",
    metadata: Mapping[str, Any] | None = None,
) -> SessionOutcome:
    _assert_outcome_kind(kind)
    _assert_outcome_status(status)
    return {
        "outcome_id": outcome_id or _derive_outcome_id(kind=kind, ref=ref, sha=sha, path=path, url=url, title=title),
        "session_id": session_id,
        "kind": kind,
        "status": status,
        "title": title or _label_for_kind(kind),
        "created_at": created_at,
        "url": url,
        "path": path,
        "ref": ref,
        "sha": sha,
        "summary": summary,
        "metadata": _sanitize_metadata(metadata or {}),
    }


def build_missing_host_capability_outcome(
    *,
    session_id: str,
    kind: SessionOutcomeKind,
    required_capability: str,
    created_at: str,
) -> SessionOutcome:
    _assert_outcome_kind(kind)
    return {
        "outcome_id": f"{kind}:missing:{required_capability}",
        "session_id": session_id,
        "kind": kind,
        "status": "unavailable",
        "title": f"{_label_for_kind(kind)} unavailable",
        "created_at": created_at,
        "url": "",
        "path": "",
        "ref": "",
        "sha": "",
        "summary": f"Host capability {required_capability} is unavailable for {kind} outcomes.",
        "metadata": {
            "reason": "missing_host_capability",
            "required_capability": required_capability,
        },
    }


def session_outcome_to_artifact(outcome: Mapping[str, Any]) -> BackgroundSessionArtifact:
    _assert_available_outcome(outcome, "artifacts")
    return {
        "artifact_id": _read_str(outcome, "outcome_id"),
        "kind": _read_str(outcome, "kind"),
        "label": _read_str(outcome, "title"),
        "path": _read_str(outcome, "path"),
        "url": _read_str(outcome, "url"),
    }


def build_session_outcome_artifact_event(
    outcome: Mapping[str, Any],
    *,
    sequence: int,
    timestamp: str,
) -> NormalizedSessionEvent:
    _assert_available_outcome(outcome, "artifact events")
    return build_artifact_created_session_event(
        session_id=_read_str(outcome, "session_id"),
        sequence=sequence,
        timestamp=timestamp,
        artifact_id=_read_str(outcome, "outcome_id"),
        kind=_read_str(outcome, "kind"),
        label=_read_str(outcome, "title") or None,
        path=_read_str(outcome, "path") or None,
        url=_read_str(outcome, "url") or None,
    )


def _derive_outcome_id(
    *,
    kind: SessionOutcomeKind,
    ref: str,
    sha: str,
    path: str,
    url: str,
    title: str,
) -> str:
    identity = ref or sha or path or url or title or kind
    return f"{kind}:{quote(identity, safe='')}"


def _sanitize_metadata(record: Mapping[str, Any]) -> dict[str, SessionOutcomeMetadataValue]:
    clean: dict[str, SessionOutcomeMetadataValue] = {}
    for key, value in record.items():
        if not isinstance(key, str) or _is_sensitive_key(key):
            continue
        if isinstance(value, str | int | float | bool):
            clean[key] = value
    return clean


def _is_sensitive_key(key: str) -> bool:
    normalized = key.lower()
    return any(
        marker in normalized
        for marker in ("secret", "token", "password", "credential", "api_key", "apikey", "private_key")
    )


def _read_str(record: Mapping[str, Any], key: str) -> str:
    value = record.get(key)
    return value if isinstance(value, str) else ""


def _label_for_kind(kind: SessionOutcomeKind) -> str:
    if kind == "branch":
        return "Branch"
    if kind == "commit":
        return "Commit"
    if kind == "pull_request":
        return "Pull request"
    if kind == "screenshot":
        return "Screenshot"
    if kind == "report":
        return "Report"
    if kind == "trace":
        return "Trace"
    if kind == "dataset":
        return "Dataset"
    return "Verification result"


def _assert_available_outcome(outcome: Mapping[str, Any], target: str) -> None:
    kind = _read_str(outcome, "kind")
    status = _read_str(outcome, "status")
    _assert_outcome_kind(kind)
    _assert_outcome_status(status)
    if status != "available":
        raise ValueError(f"Only available session outcomes can be converted to {target}")


def _assert_outcome_kind(kind: str) -> None:
    if kind not in _OUTCOME_KINDS:
        raise ValueError(f"Unsupported session outcome kind: {kind}")


def _assert_outcome_status(status: str) -> None:
    if status not in _OUTCOME_STATUSES:
        raise ValueError(f"Unsupported session outcome status: {status}")

File: ts/src/session/background-session-outcomes.ts
import { buildArtifactCreatedSessionEvent, type NormalizedSessionEvent } from "./background-session-events.js";
import type {
  BackgroundSessionArtifact,
  BackgroundSessionArtifactKind,
} from "./background-session-read-model.js";

export type SessionOutcomeKind =
  | "branch"
  | "commit"
  | "pull_request"
  | "screenshot"
  | "report"
  | "trace"
  | "dataset"
  | "verification_result";

export type SessionOutcomeStatus = "available" | "pending" | "unavailable";
export type SessionOutcomeMetadataValue = string | number | boolean;

export interface SessionOutcomeInput {
  readonly sessionId: string;
  readonly kind: SessionOutcomeKind;
  readonly outcomeId?: string;
  readonly status?: SessionOutcomeStatus;
  readonly title?: string;
  readonly createdAt: string;
  readonly url?: string;
  readonly path?: string;
  readonly ref?: string;
  readonly sha?: string;
  readonly summary?: string;
  readonly metadata?: Record<string, unknown>;
}

export interface SessionOutcome {
  readonly outcome_id: string;
  readonly session_id: string;
  readonly kind: SessionOutcomeKind;
  readonly status: SessionOutcomeStatus;
  readonly title: string;
  readonly created_at: string;
  readonly url: string;
  readonly path: string;
  readonly ref: string;
  readonly sha: string;
  readonly summary: string;
  readonly metadata: Record<string, SessionOutcomeMetadataValue>;
}

export interface MissingHostCapabilityOutcomeInput {
  readonly sessionId: string;
  readonly kind: SessionOutcomeKind;
  readonly requiredCapability: string;
  readonly createdAt: string;
}

export interface SessionOutcomeArtifactEventInput {
  readonly sequence: number;
  readonly timestamp: string;
}

export function buildSessionOutcome(input: SessionOutcomeInput): SessionOutcome {
  assertSessionOutcomeKind(input.kind);
  const status = input.status ?? "available";
  assertSessionOutcomeStatus(status);
  const normalized: SessionOutcome = {
    outcome_id: input.outcomeId || deriveOutcomeId(input),
    session_id: input.sessionId,
    kind: input.kind,
    status,
    title: input.title || labelForKind(input.kind),
    created_at: input.createdAt,
    url: input.url ?? "",
    path: input.path ?? "",
    ref: input.ref ?? "",
    sha: input.sha ?? "",
    summary: input.summary ?? "",
    metadata: sanitizeMetadata(input.metadata ?? {}),
  };
  return normalized;
}

export function buildMissingHostCapabilityOutcome(
  input: MissingHostCapabilityOutcomeInput,
): SessionOutcome {
  assertSessionOutcomeKind(input.kind);
  return {
    outcome_id: `${input.kind}:missing:${input.requiredCapability}`,
    session_id: input.sessionId,
    kind: input.kind,
    status: "unavailable",
    title: `${labelForKind(input.kind)} unavailable`,
    created_at: input.createdAt,
    url: "",
    path: "",
    ref: "",
    sha: "",
    summary: `Host capability ${input.requiredCapability} is unavailable for ${input.kind} outcomes.`,
    metadata: {
      reason: "missing_host_capability",
      required_capability: input.requiredCapability,
    },
  };
}

export function sessionOutcomeToArtifact(outcome: SessionOutcome): BackgroundSessionArtifact {
  assertAvailableSessionOutcome(outcome, "artifacts");
  return {
    artifact_id: outcome.outcome_id,
    kind: outcome.kind as BackgroundSessionArtifactKind,
    label: outcome.title,
    path: outcome.path,
    url: outcome.url,
  };
}

export function buildSessionOutcomeArtifactEvent(
  outcome: SessionOutcome,
  input: SessionOutcomeArtifactEventInput,
): NormalizedSessionEvent {
  assertAvailableSessionOutcome(outcome, "artifact events");
  return buildArtifactCreatedSessionEvent({
    sessionId: outcome.session_id,
    sequence: input.sequence,
    timestamp: input.timestamp,
    artifactId: outcome.outcome_id,
    kind: outcome.kind,
    label: outcome.title || undefined,
    path: outcome.path || undefined,
    url: outcome.url || undefined,
  });
}

function deriveOutcomeId(input: SessionOutcomeInput): string {
  const identity = input.ref || input.sha || input.path || input.url || input.title || input.kind;
  return `${input.kind}:${encodeURIComponent(identity)}`;
}

function sanitizeMetadata(record: Record<string, unknown>): Record<string, SessionOutcomeMetadataValue> {
  const clean: Record<string, SessionOutcomeMetadataValue> = {};
  for (const [key, value] of Object.entries(record)) {
    if (isSensitiveKey(key)) {
      continue;
    }
    if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
      clean[key] = value;
    }
  }
  return clean;
}

function isSensitiveKey(key: string): boolean {
  const normalized = key.toLowerCase();
  return ["secret", "token", "password", "credential", "api_key", "apikey", "private_key"].some(
    (marker) => normalized.includes(marker),
  );
}

function assertAvailableSessionOutcome(outcome: SessionOutcome, target: string): void {
  assertSessionOutcomeKind(outcome.kind);
  assertSessionOutcomeStatus(outcome.status);
  if (outcome.status !== "available") {
    throw new RangeError(`Only available session outcomes can be converted to ${target}`);
  }
}

const SESSION_OUTCOME_KINDS = new Set<string>([
  "branch",
  "commit",
  "pull_request",
  "screenshot",
  "report",
  "trace",
  "dataset",
  "verification_result",
]);

function assertSessionOutcomeKind(kind: unknown): asserts kind is SessionOutcomeKind {
  if (typeof kind !== "string" || !SESSION_OUTCOME_KINDS.has(kind)) {
    throw new RangeError(`Unsupported session outcome kind: ${String(kind)}`);
  }
}

const SESSION_OUTCOME_STATUSES = new Set<string>(["available", "pending", "unavailable"]);

function assertSessionOutcomeStatus(status: unknown): asserts status is SessionOutcomeStatus {
  if (typeof status !== "string" || !SESSION_OUTCOME_STATUSES.has(status)) {
    throw new RangeError(`Unsupported session outcome status: ${String(status)}`);
  }
}

function labelForKind(kind: SessionOutcomeKind): string {
  switch (kind) {
    case "branch":
      return "Branch";
    case "commit":
      return "Commit";
    case "pull_request":
      return "Pull request";
    case "screenshot":
      return "Screenshot";
    case "report":
      return "Report";
    case "trace":
      return "Trace";
    case "dataset":
      return "Dataset";
    case "verification_result":
      return "Verification result";
  }
}
